ibis/expr/analysis.py

```diff
@@ -220,22 +221,13 @@ def find_immediate_parent_tables(expr):
             Literal[int8]
               1
     """
-    stack = [expr]
-    seen = set()
-
-    while stack:
-        e = stack.pop()
-        node = e.op()
-
-        if node not in seen:
-            seen.add(node)
+    def finder(expr):
+        if isinstance(expr, ir.TableExpr):
+            return lin.stop, expr
+        else:
+            return lin.proceed, None
```

Comment:
```python
def finder(expr):
    if isinstance(expr, ir.TableExpr):
        return expr

def proceed(expr, result):
    return not isinstance(expr, ir.TableExpr)

lin.traverse(finder, expr, proceed=proceed)
```

Reply:
Yep, that looks good to me.

Reply:
Ehh, You need to convince me about the advantages :)
I favor the previouse one.

Reply:
In case of [lineage](https://github.com/ibis-project/ibis/blob/36251a1e170d3126023897e77f6d45f0714f91d4/ibis/expr/lineage.py#L167) function the callable must return the next traversable expressions.